src/main/java/com/gtnewhorizon/gtnhlib/capability/item/ItemTransfer.java

```diff
@@ -0,0 +1,179 @@
+    @Setter
+    protected int stacksToTransfer = 1;
+    @Setter
+    protected int maxItemsPerTransfer = 64;
+    @Setter
+    protected int maxTotalTransferred = Integer.MAX_VALUE;
+
+    @Getter
+    protected int totalItemsTransferred = 0;
+    @Getter
+    protected int totalStacksTransferred = 0;
+    @Getter
+    protected int prevItemsTransferred = 0;
+    @Getter
+    protected int prevStacksTransferred = 0;
+
+    protected int[] sourceSlots, sinkSlots;
+
+    @Setter
+    protected ItemStackPredicate filter;
+
+    public void source(IItemSource source) {
+        this.source = source;
+    }
+
+    public void source(Object source, ForgeDirection side) {
+        this.source = ItemUtil.getItemSource(source, side, true);
+    }
+
+    public void sink(IItemSink sink) {
+        this.sink = sink;
+    }
+
+    public void sink(Object sink, ForgeDirection side) {
+        this.sink = ItemUtil.getItemSink(sink, side, true);
+    }
+
+    public <Coord extends IBlockPos & IWorldReferent> void push(Coord pos, ForgeDirection side) {
+        TileEntity self = pos.getWorld().getTileEntity(pos.getX(), pos.getY(), pos.getZ());
+        TileEntity adjacent = pos.getWorld()
+                .getTileEntity(pos.getX() + side.offsetX, pos.getY() + side.offsetY, pos.getZ() + side.offsetZ);
+
+        push(self, side, adjacent);
+    }
+
+    public void push(Object self, ForgeDirection side, Object target) {
+        source(self, side);
+        sink(target, side.getOpposite());
+    }
+
+    public <Coord extends IBlockPos & IWorldReferent> void pull(Coord pos, ForgeDirection side) {
+        TileEntity self = pos.getWorld().getTileEntity(pos.getX(), pos.getY(), pos.getZ());
+        TileEntity adjacent = pos.getWorld()
+                .getTileEntity(pos.getX() + side.offsetX, pos.getY() + side.offsetY, pos.getZ() + side.offsetZ);
+
+        pull(self, side, adjacent);
+    }
+
+    public void pull(Object self, ForgeDirection side, Object target) {
+        source(target, side.getOpposite());
+        sink(self, side);
+    }
+
+    public void dropItems(World world, IBlockPos pos) {
+        this.itemDropping = new DroppingItemSink(world, pos);
+    }
+
+    public <Coord extends IBlockPos & IWorldReferent> void dropItems(Coord pos, ForgeDirection output) {
+        dropItems(pos.getWorld(), pos.offset(output));
+    }
+
+    public void setSourceSlots(int... sourceSlots) {
+        this.sourceSlots = sourceSlots;
+    }
+
+    public void setSinkSlots(int... sinkSlots) {
+        this.sinkSlots = sinkSlots;
+    }
+
+    public int transfer() {
+        IItemSink sink = this.sink == null ? itemDropping : this.sink;
+
+        if (source == null) return 0;
+        if (sink == null) return 0;
+        if (stacksToTransfer == 0) return 0;
+        if (maxItemsPerTransfer == 0) return 0;
+
+        source.setAllowedSourceSlots(sourceSlots);
+        sink.setAllowedSinkSlots(sinkSlots);
+
+        InventorySourceIterator iter = source.iterator();
```

Comment:
Why does this function use `InventorySourceIterator`? Shouldn't the Source/Sink be enough? The sink does so much work finding the correct slots to put stuff in, why can't  the source have a single extract method that finds the next matching itemstack?